src/group_theory/noncomm_pi_coprod.lean

```diff
@@ -0,0 +1,478 @@
+  `⨆ (i : ι), (ϕ i).mrange`
+* `monoid_hom.noncomm_pi_coprod_range`: The range of `monoid_hom.noncomm_pi_coprod` is
+  `⨆ (i : ι), (ϕ i).range`
+* `subgroup.noncomm_pi_coprod_range`: The range of `subgroup.noncomm_pi_coprod` is `⨆ (i : ι), H i`.
+* `monoid_hom.injective_noncomm_pi_coprod_of_independent`: in the case of groups, `pi_hom.hom` is
+   injective if the `ϕ` are injective and the ranges of the `ϕ` are independent.
+* `monoid_hom.independent_range_of_coprime_order`: If the `N i` have coprime orders, then the ranges
+   of the `ϕ` are independent.
+* `subgroup.independent_of_coprime_order`: If commuting, normal subgroups `H i` have coprime orders,
+   they are independent.
+
+-/
+
+open_locale big_operators
+
+-- I think it's worth keeping it and moving to appropriate file
+@[to_additive]
+lemma mul_eq_one_iff_disjoint {G : Type*} [group G] {H₁ H₂ : subgroup G} :
+  disjoint H₁ H₂ ↔ ∀ {x y : G}, x ∈ H₁ → y ∈ H₂ → x * y = 1 → x = 1 ∧ y = 1 :=
+begin
+  split,
+  { intros hdis x y hx hy heq,
+    obtain rfl : y = x⁻¹ := symm (inv_eq_iff_mul_eq_one.mpr heq),
+    have hy := H₂.inv_mem_iff.mp hy,
+    have : x ∈ H₁ ⊓ H₂, by { simp, cc },
+    rw [hdis.eq_bot, subgroup.mem_bot] at this,
+    subst this,
+    simp },
+  { rintros h x ⟨hx1, hx2⟩,
+    obtain rfl : x = 1 := (h hx1 (H₂.inv_mem hx2) (mul_inv_self x)).1,
+    exact rfl, },
+end
+
+section family_of_monoids
+
+variables {M : Type*} [monoid M]
+
+-- We have a family of monoids
+-- The fintype assumption is not always used, but declared here, to keep things in order
+variables {ι : Type*} [hdec : decidable_eq ι] [hfin : fintype ι]
+variables {N : ι → Type*} [∀ i, monoid (N i)]
+
+@[to_additive]
+lemma _root_.finset.noncomm_prod_map {α : Type*} {β : Type*} [monoid β] (s : finset α) (f : α → β)
+  (comm : ∀ (x : α), x ∈ s → ∀ (y : α), y ∈ s → commute (f x) (f y))
+  {M : Type*} [monoid M] (g : β →* M) :
+  g (s.noncomm_prod f comm) = s.noncomm_prod (λ i, g (f i))
+  (λ x hx y hy, commute.map (comm x hx y hy) g)  :=
+sorry
+
+@[to_additive]
+lemma _root_.finset.noncomm_prod_eq_one {α : Type*} {β : Type*} [monoid β] (s : finset α)
+  (f : α → β) (comm : ∀ (x : α), x ∈ s → ∀ (y : α), y ∈ s → commute (f x) (f y))
+  (h : ∀ (x : α), x ∈ s → f x = 1) : s.noncomm_prod f comm = 1 :=
+sorry
+
+@[to_additive]
+lemma _root_.finset.noncomm_prod_single [decidable_eq ι] [fintype ι] (x : Π i, N i) :
+  finset.univ.noncomm_prod (λ i, monoid_hom.single N i (x i))
+  (λ i _ j _, by { by_cases h : i = j, { subst h, }, { exact pi.mul_single_commute i j h _ _ } })
+  = x :=
+begin
+  ext i,
+  apply (finset.univ.noncomm_prod_map (λ i, monoid_hom.single N i (x i)) _
+    (pi.eval_monoid_hom _ i)).trans,
+  rw (finset.insert_erase (finset.mem_univ i)).symm,
+  rw finset.noncomm_prod_insert_of_not_mem',
+  rw finset.noncomm_prod_eq_one,
+  { simp, },
+  { intros i h,  simp at h, simp [h], },
+  { simp, },
+end
+
+@[to_additive]
+lemma _root_.monoid_hom.pi_ext [decidable_eq ι] [fintype ι] {f g : (Π i, N i) →* M}
```

Comment:
I think it's best to use
https://leanprover-community.github.io/mathlib_docs/group_theory/submonoid/basic.html#monoid_hom.eq_of_eq_on_mdense
combined with
`lemma mrange : (hom ϕ hcomm S).mrange = ⨆ i ∈ S, (ϕ i).mrange`
applied to the inclusions of `H i` into the pi group (I think the ⨆ is defined as an Inf).
But if mrange is used then this can't be moved to the group.pi ... Finiteness of ι is also necessary here.

Reply:
Hmm, I’m not sure if [`monoid_hom.eq_of_eq_on_mdense`](https://leanprover-community.github.io/mathlib_docs/group_theory/submonoid/basic.html#monoid_hom.eq_of_eq_on_mdense) saves me a lot – proof I have is already  quite short, building on a lemma stating that an element of `(x : Π i, N i)` is the product of it’s components if `ι` is finite. And that I have to prove anyways, whether in the form I have (as an explicit equality with `noncomm_prod`), or somehow phrased as saying that in  `(x : Π i, N i)` the closure of the `mrange`s of the `monoid_hom.single` is the whole group.

Maybe all of this should go into a new file `algebra.group.pi_finite`?

Reply:
Indeed, going through mdense seems convoluted, and your approach is more straightforward. `noncomm_prod_single` is the key lemma, and I'm now trying to fill in the `sorry`s by converting `noncomm_prod` to `list.prod`.

Reply:
Sorries killed! I think the first two lemmas could go into `noncomm_prod` while the other two could go into `group.pi` (importing `noncomm_prod`). I don't think it's worth creating a separate file for the finite case; when people have developed a multiplicative version of `finsupp` they may go back to generalize the current development to arbitrary index type.

Reply:
Thanks! I generalized them a bit and will now move them into appropriate files and create a second PR.

Reply:
> while the other two could go into group.pi (importing noncomm_prod)

It seems that `noncomm_prod` already imports `group.pi` indirectly, so I think they’ll go there as well

Reply:
Ok, the yak has grown a few hairs again, to be shaved at https://github.com/leanprover-community/mathlib/pull/12291. I hope we don’t keep growing PRs at a fast rate than we can merge them :-)